### PR TITLE
UCP/CORE: Don't destroy UCT EP from UCT pending callback

### PR DESCRIPTION
**Pull request by dmitrygx**
## What

Don't destroy UCT EP from UCT pending callback.

## Why ?

To avoid possible arbiter guard check failures, when `uct_ep_flush()` returns OK in a pending callback which is called from an arbiter. It causes calling `ucs_arbiter_group_purge()` from `uct_ep_destroy()` while UCT is in `ucs_arbiter_dispatch()`
```
[<host>:9399 :0:9410]     arbiter.c:116  Assertion `(group)->guard == 0' failed: scheduling arbiter group 0x7fffbc161ae0 while it's being dispatched

/labhome/dmitrygla/work_auto/ucx/src/ucs/datastruct/arbiter.c: [ ucs_arbiter_group_purge() ]
      ...
      113         return; /* Empty group */
      114     }
      115
==>   116     UCS_ARBITER_GROUP_GUARD_CHECK(group);
      117
      118     head = tail->next;
      119     next = head;

==== backtrace (tid:   9410) ====
 0 0x0000000000050d22 ucs_arbiter_group_purge()  /labhome/dmitrygla/work_auto/ucx/src/ucs/datastruct/arbiter.c:116
 1 0x0000000000124c93 uct_dc_mlx5_ep_pending_purge()  /labhome/dmitrygla/work_auto/ucx/src/uct/ib/dc/dc_mlx5_ep.c:1337
 2 0x00000000001237c8 uct_dc_mlx5_ep_t_cleanup()  /labhome/dmitrygla/work_auto/ucx/src/uct/ib/dc/dc_mlx5_ep.c:1036
 3 0x0000000000079cf2 ucs_class_call_cleanup_chain()  /labhome/dmitrygla/work_auto/ucx/src/ucs/type/class.c:56
 4 0x0000000000123c38 uct_dc_mlx5_ep_t_delete()  /labhome/dmitrygla/work_auto/ucx/src/uct/ib/dc/dc_mlx5_ep.c:1072
 5 0x0000000000014151 uct_ep_destroy()  /labhome/dmitrygla/work_auto/ucx/src/uct/base/uct_iface.c:478
 6 0x0000000000056051 ucp_worker_discard_uct_ep_destroy_progress()  /labhome/dmitrygla/work_auto/ucx/src/ucp/core/ucp_worker.c:2218
 7 0x00000000000562e0 ucp_worker_discard_uct_ep_pending_cb()  /labhome/dmitrygla/work_auto/ucx/src/ucp/core/ucp_worker.c:2259
 8 0x0000000000124749 uct_rc_iface_invoke_pending_cb()  /labhome/dmitrygla/work_auto/ucx/src/uct/ib/rc/base/rc_iface.h:526
 9 0x0000000000124749 uct_dc_mlx5_iface_dci_do_common_pending_tx()  /labhome/dmitrygla/work_auto/ucx/src/uct/ib/dc/dc_mlx5_ep.c:1212
10 0x00000000001248bb uct_dc_mlx5_iface_dci_do_dcs_pending_tx()  /labhome/dmitrygla/work_auto/ucx/src/uct/ib/dc/dc_mlx5_ep.c:1245
11 0x00000000000514fb ucs_arbiter_dispatch_nonempty()  /labhome/dmitrygla/work_auto/ucx/src/ucs/datastruct/arbiter.c:320
12 0x00000000000c4d46 ucs_arbiter_dispatch()  /labhome/dmitrygla/work_auto/ucx/src/ucs/datastruct/arbiter.h:386
13 0x00000000001256b8 uct_dc_mlx5_iface_progress_pending()  /labhome/dmitrygla/work_auto/ucx/src/uct/ib/dc/dc_mlx5_ep.h:312
14 0x00000000001256b8 uct_dc_mlx5_ep_handle_failure()  /labhome/dmitrygla/work_auto/ucx/src/uct/ib/dc/dc_mlx5_ep.c:1386
15 0x0000000000147dcb uct_dc_mlx5_dci_handle_failure()  /labhome/dmitrygla/work_auto/ucx/src/uct/ib/dc/dc_mlx5.c:1081
16 0x0000000000147e64 uct_dc_mlx5_iface_handle_failure()  /labhome/dmitrygla/work_auto/ucx/src/uct/ib/dc/dc_mlx5.c:1096
17 0x0000000000029963 uct_ib_mlx5_check_completion()  /labhome/dmitrygla/work_auto/ucx/src/uct/ib/mlx5/ib_mlx5.c:360
18 0x000000000013bb8a uct_ib_mlx5_poll_cq()  /labhome/dmitrygla/work_auto/ucx/src/uct/ib/mlx5/ib_mlx5.inl:81
19 0x000000000013bb8a uct_dc_mlx5_poll_tx()  /labhome/dmitrygla/work_auto/ucx/src/uct/ib/dc/dc_mlx5.c:207
20 0x000000000013bb8a uct_dc_mlx5_iface_progress()  /labhome/dmitrygla/work_auto/ucx/src/uct/ib/dc/dc_mlx5.c:252
21 0x000000000013bb8a uct_dc_mlx5_iface_progress_ll()  /labhome/dmitrygla/work_auto/ucx/src/uct/ib/dc/dc_mlx5.c:262
22 0x000000000004b33c ucs_callbackq_dispatch()  /labhome/dmitrygla/work_auto/ucx/src/ucs/datastruct/callbackq.h:211
23 0x0000000000056ec7 uct_worker_progress()  /labhome/dmitrygla/work_auto/ucx/src/uct/api/uct.h:2436
...
```

## How ?

If `uct_ep_flush()` returns `UCS_OK`, schedule `ucp_worker_discard_uct_ep_destroy_progress()` to destroy UCT EP from the progress.

### CODE REVIEW

**Review by yosefe — APPROVED** [review]
nice catch! how was it found?

**Review by brminich — APPROVED** [review]

**Comment by dmitrygx** [comment]
> nice catch! how was it found?

@yosefe it was found during implementing another fix

**Review by evgeny-leksikov — APPROVED** [review]
👍

File: src/ucp/wireup/wireup_cm.c
Comment (evgeny-leksikov) [inline src/ucp/wireup/wireup_cm.c:763]: BTW, unrelated but should not we do async block here?

Comment (dmitrygx) [reply src/ucp/wireup/wireup_cm.c:763]: we should - it is done in #6312

**Comment by dmitrygx** [comment]
@evgeny-leksikov too many places where we depend on the fact that CM lane is discarded in-place. So, better to fix it by not adding on-progress if it is not called from pending_cb

**Comment by evgeny-leksikov** [comment]
> @evgeny-leksikov too many places where we depend on the fact that CM lane is discarded in-place. So, better to fix it by not adding on-progress if it is not called from pending_cb

IMO this is bad assumption

**Review by evgeny-leksikov — CHANGES_REQUESTED** [review]
rescheduling any event in progress queue should not cause failures
pls do not force push until everything is OK

**Comment by dmitrygx** [comment]
> rescheduling any event in progress queue should not cause failures

yes, I agree. let's have the previous implementation but with additional fixes
also, we should then update discarding to not put into the progress when it is not really needed

File: src/ucp/wireup/wireup_cm.c
Comment (dmitrygx) [reply src/ucp/wireup/wireup_cm.c]: fixed

File: src/ucp/wireup/wireup.c
Comment (dmitrygx) [reply src/ucp/wireup/wireup.c]: fixed

File: src/ucp/core/ucp_ep.inl
Comment (dmitrygx) [reply src/ucp/core/ucp_ep.inl]: done

File: src/ucp/core/ucp_ep.c
Comment (dmitrygx) [reply src/ucp/core/ucp_ep.c]: fixed in another PR which is already merged

File: src/ucp/core/ucp_ep.c
Comment (dmitrygx) [reply src/ucp/core/ucp_ep.c]: removed

**Comment by dmitrygx** [comment]
@evgeny-leksikov @brminich could you review pls?

File: src/ucp/wireup/wireup_cm.c
Comment (evgeny-leksikov) [inline src/ucp/wireup/wireup_cm.c:329]: seems like the error code does not reflect actual one, we had status in EP config key for such cases and probably deleted this...

Comment (dmitrygx) [reply src/ucp/wireup/wireup_cm.c:329]: yes, we deleted it, but we set nothing to the `status`, it was UCS_OK always.

Comment (dmitrygx) [reply src/ucp/wireup/wireup_cm.c:329]: also, I think `status` should be stored in UCP control extension rather in EP config key
`status` belongs to EP, but not to all EPs which use this EP config key

Comment (evgeny-leksikov) [reply src/ucp/wireup/wireup_cm.c:329]: the ep assumed to be reconfigured on failure with relevant status, we don't expect a lot of reasons of failure and all failed EPs with the same status should have the same configuration

Comment (yosefe) [reply src/ucp/wireup/wireup_cm.c:329]: - assert that if FAILED, also CLOSED
- return ERR_CANCELED

**Comment by yosefe** [comment]
bot:pipe:retest

**Comment by dmitrygx** [comment]
@evgeny-leksikov @yosefe @brminich could you review pls?

File: src/ucp/core/ucp_worker.c
Comment (yosefe) [inline src/ucp/core/ucp_worker.c:2322]: is this a fix or enhancement??

File: src/ucp/core/ucp_worker.c
Comment (yosefe) [inline src/ucp/core/ucp_worker.c:2342]: what would happen if we removed this check and did flush-cancel regardless of err mode?
(pls add this to the comment)

File: src/ucp/wireup/wireup.c
Comment (yosefe) [inline src/ucp/wireup/wireup.c:682]: is this needed because of https://github.com/openucx/ucx/pull/6401/files#diff-1a6cd9df9fc4c2ac2f3e662ca4ce0ff54e0ad405bab393e80090c8593745351cR2342?

File: src/ucp/wireup/wireup_cm.c
Comment (yosefe) [inline src/ucp/wireup/wireup_cm.c:25]: 1. UCP_EP_CM_CALLBACK_ENTER(_ucp_ep, _uct_cm_ep, _failed_action)
2. The doc is unclear/confusing. Pls replace by just a short doc of macro args

File: src/ucp/wireup/wireup_cm.c
Comment (yosefe) [inline src/ucp/wireup/wireup_cm.c:354]: seems wrong. what if we have address/route error?

File: src/ucp/wireup/wireup_cm.c
Comment (yosefe) [inline src/ucp/wireup/wireup_cm.c:646]: why is it ok to check ep flags w/o async lock?

File: src/ucp/wireup/wireup_cm.c
Comment (yosefe) [inline src/ucp/wireup/wireup_cm.c:824]: why is it ok to check ep flags w/o async lock?

File: src/ucp/wireup/wireup_cm.c
Comment (dmitrygx) [reply src/ucp/wireup/wireup_cm.c:646]: added check that async lock is blocked when checking this flag

File: src/ucp/wireup/wireup_cm.c
Comment (dmitrygx) [reply src/ucp/wireup/wireup_cm.c:824]: added check that async lock is blocked when checking this flag

File: src/ucp/wireup/wireup_cm.c
Comment (yosefe) [inline src/ucp/wireup/wireup_cm.c:30]: 👍

File: src/ucp/wireup/wireup_cm.c
Comment (yosefe) [inline src/ucp/wireup/wireup_cm.c:25]: minor: can remove some spaces after [in]

File: src/ucp/wireup/wireup_cm.c
Comment (yosefe) [inline src/ucp/wireup/wireup_cm.c:650]: are we relying on connect_cb being called only from the async thread?

**Review by yosefe — APPROVED** [review]
pls squash

File: src/ucp/wireup/wireup_cm.c
Comment (dmitrygx) [reply src/ucp/wireup/wireup_cm.c:650]: yes, we rely on this

File: src/ucp/wireup/wireup_cm.c
Comment (dmitrygx) [reply src/ucp/wireup/wireup_cm.c]: fixed

File: src/ucp/wireup/wireup_cm.c
Comment (brminich) [inline src/ucp/wireup/wireup_cm.c:21]: `CM`

Comment (dmitrygx) [reply src/ucp/wireup/wireup_cm.c]: fixed

**Review by brminich — APPROVED** [review]

File: src/ucp/wireup/wireup_cm.c
Comment (evgeny-leksikov) [reply src/ucp/wireup/wireup_cm.c:650]: seems like this is wrong
https://github.com/openucx/ucx/blob/fc1c37f131bcb8c6d27991c228dbb4a5e4b3b206/src/uct/api/uct.h#L559

Comment (evgeny-leksikov) [reply src/ucp/wireup/wireup_cm.c:650]: for any async callback

**Comment by dmitrygx** [comment]
bot:pipe:retest

File: src/ucp/wireup/wireup_cm.c
Comment (dmitrygx) [reply src/ucp/wireup/wireup_cm.c:650]: discussed offline
will do ASYNC block/unblock in CM callbacks in the separate PR

**Comment by dmitrygx** [comment]
the 2nd commit fixes #6543 - we should purge pending before destroying UCP EP in UCP worker destroy

File: src/ucp/core/ucp_worker.c
Comment (brminich) [inline src/ucp/core/ucp_worker.c:2350]: minor: space before `*/`

File: src/ucp/core/ucp_worker.c
Comment (brminich) [inline src/ucp/core/ucp_worker.c:2342]: what if error handling mode is NONE, but pendings are present?

Comment (dmitrygx) [reply src/ucp/core/ucp_worker.c:2342]: it will be failed with fatal error in the callback - and printing a message in `ucs_bug()`

File: src/ucp/core/ucp_worker.c
Comment (dmitrygx) [reply src/ucp/core/ucp_worker.c]: fixed

File: src/ucp/core/ucp_worker.c
Comment (dmitrygx) [reply src/ucp/core/ucp_worker.c:2342]: I think we don't need to take care of such EPs. Normally, user's EPs should not go here.

Comment (brminich) [reply src/ucp/core/ucp_worker.c:2342]: so user is not allowed to destroy worker until all eps are flushed/closed?

Comment (evgeny-leksikov) [reply src/ucp/core/ucp_worker.c:2342]: that is not allowed for any err_mode

Comment (dmitrygx) [reply src/ucp/core/ucp_worker.c:2342]: yes, but anyway if error handling was requested, we can purge pending/outstanding operations to a user

Comment (yosefe) [reply src/ucp/core/ucp_worker.c:2342]: @dmitrygx didn't we say it's no longer needed after UD fixes?
what happens if we remove the check?

Comment (dmitrygx) [reply src/ucp/core/ucp_worker.c:2342]: > @dmitrygx didn't we say it's no longer needed after UD fixes?

unfortunately, I found that UD moves purging outstanding operations from UD iface to EP, but if the operations are pending, so these requests will be purged with a callback that calls `ucs_bug()`

> what happens if we remove the check?

crash with XPMEM (it touches remote peer memory which couldn't be available anymore) which doesn't support error handling and hence flush(CANCEL)

Comment (evgeny-leksikov) [reply src/ucp/core/ucp_worker.c:2342]: should not workers to be flushed before destroy if err handling was not requested?

Comment (dmitrygx) [reply src/ucp/core/ucp_worker.c]: updated the PR to implement and use `ucp_ep_purge_lanes()` instead of `ucp_ep_discard_lanes()` which does unnecessary action - flush(CANCEL).
btw, there are two places where we do purge of pending request in a loop for all EP lanes, but also it does some other actions in the same loop. Shall we split this loop into two loops: 1) pending purging, 2) other actions, e.g. destroying UCT EP? The 1) action will done by calling `ucp_ep_purge_lanes()`. wdyt?

Comment (dmitrygx) [reply src/ucp/core/ucp_worker.c]: > should not workers to be flushed before destroy if err handling was not requested?

Yes, we should do it. But we don't wait for Reply EP being completed its work, e.g. by incrementing/decrementing flush_ops counter in UCP Worker

File: src/ucp/core/ucp_worker.c
Comment (yosefe) [inline src/ucp/core/ucp_worker.c:2344]: what is the problem with doing it from ucp_destroyed_ep_pending_purge?
maybe we should replace ucp_destroyed_ep_pending_purge by ucp_ep_err_pending_purge?

Comment (dmitrygx) [reply src/ucp/core/ucp_worker.c:2344]: `ucp_destroyed_ep_pending_purge()` was added to report that a user's requests weren't sent correctly, but EP was closed
so, removing this, we won't be able to catch this anymore

Comment (yosefe) [reply src/ucp/core/ucp_worker.c:2344]: so should it be like
```
/* Cleanup pending operations on the endpoint, since ucp_ep_destroy_internal() expects the pending
queues of the UCT endpoints will be empty before they are destroyed */
```

File: src/ucp/core/ucp_worker.c
Comment (evgeny-leksikov) [reply src/ucp/core/ucp_worker.c]: why? IMO all hidden eps (reply, KA, etc) should part of worker flush as well

Comment (dmitrygx) [reply src/ucp/core/ucp_worker.c]: > why? IMO all hidden eps (reply, KA, etc) should part of worker flush as well

yes, if we do flush all EPs as well

File: src/ucp/core/ucp_worker.c
Comment (dmitrygx) [reply src/ucp/core/ucp_worker.c:2344]: yes

File: src/ucp/core/ucp_worker.c
Comment (yosefe) [reply src/ucp/core/ucp_worker.c]: we don't require flushing the worker before destroying it

File: src/ucp/core/ucp_worker.c
Comment (dmitrygx) [reply src/ucp/core/ucp_worker.c]: done

**Review by yosefe — APPROVED** [review]

**Comment by dmitrygx** [comment]
squashed - waiting for CI
crossed fingers - hope. nothing wasn't missed :)

File: src/ucp/core/ucp_worker.c
Comment (brminich) [reply src/ucp/core/ucp_worker.c]: but then we have eps  (KA) which is not handled as all others, i. e. not flushed (which requires special handling, like bug fix here).

Comment (dmitrygx) [reply src/ucp/core/ucp_worker.c]: the EP created for sending EP_REMOVED answer during KA is created only if error handling requested by user.
E.g. We don't do flush in gtests for error handling tests

**Comment by dmitrygx** [comment]
@brminich could you review pls?

File: src/ucp/core/ucp_worker.c
Comment (yosefe) [reply src/ucp/core/ucp_worker.c]: we can silently destroy such KA-reply temporary ep and cancel all its operations internally and not wait for them to complete since we are going to destroy all transports anyway

Comment (evgeny-leksikov) [reply src/ucp/core/ucp_worker.c]: AFAIU, KA-reply EPs has to be created only on UCT ifaces which support peer failure err handling (we should request it on internal ep_create(ka_ep)), so we can assert or I'm missing something?

Comment (dmitrygx) [reply src/ucp/core/ucp_worker.c]: I've removed the check here
So, now we purge pending requests for all EPs (KA-reply EPs, EPs which weren't exposed to a user, EPs which weren't destroyed y a user).

**Review by brminich — APPROVED** [review]

**Pull request merged**
